### ISSUE TITLE
Python Deep Learning Libraries

### ISSUE THREAD

**Issue by sepidehamiri** [sepidehamiri/personal-research#8]
## 1. [Caffe](http://caffe.berkeleyvision.org/)
Caffe is a **deep learning framework** developed by the Berkeley Vision and Learning Center (BVLC). It’s **modular**. Extremely fast. And it’s used by academics and industry in start-of-the-art applications. While Caffe itself isn’t a Python library, it does provide bindings into the Python programming language. We typically use these bindings when actually deploying our network in the wild. The reason I’ve included Caffe in this list is because it’s used nearly everywhere.
You define your model architecture and solver methods in a plaintext, JSON-like file called _.prototxt_  configuration files. The Caffe binaries take these _.prototxt_  files and train your network. After Caffe is done training, you can take your network and classify new images via Caffe binaries, or better yet, through the Python or MATLAB APIs. Caffe is expressive architecture encourages application and innovation. Models and optimization are defined by configuration without hard-coding. Switch between CPU and GPU by setting a single flag to train on a GPU machine then deploy to commodity clusters or mobile devices.
Speed makes Caffe perfect for research experiments and industry deployment. **Caffe can process over 60M images per day with a single NVIDIA K40 GPU***. That’s 1 ms/image for inference and 4 ms/image for learning and more recent library versions and hardware are faster still. We believe that **Caffe is among the fastest convnet implementations available.** Caffe already powers academic research projects, startup prototypes, and even large-scale industrial applications in vision, speech, and multimedia.
Problem of caffe is that constructing an architecture inside the .prototxt  files can become quite tedious and tiresome. And more to the point, tuning hyper parameters with Caffe can not be (easily) done pro grammatically! Because of these two reasons, I tend to lean towards libraries that allow me to implement the end-to-end network (including cross-validation and hyper parameter tuning) in a Python-based API.

## 2. [Theano](http://deeplearning.net/software/theano/)
Theano is a Python library that allows you to **define, optimize, and evaluate mathematical expressions involving multi-dimensional arrays efficiently**. Theano accomplishes this via tight integration with NumPy and transparent use of the GPU. While you can build deep learning networks in Theano, I tend to think of Theano as the building blocks for neural networks, in the same way that NumPy serves as the building blocks for scientific computing. While not a perfect comparison, building a Convolutional Neural Network in Theano is like writing a custom Support Vector Machine (SVM) in native Python with only a sprinkle of NumPy.
Theano features:
**tight integration with NumPy** – Use numpy.ndarray in Theano-compiled functions.
**transparent use of a GPU** – Perform data-intensive computations much faster than on a CPU.
**efficient symbolic differentiation** – Theano does your derivatives for functions with one or many inputs.
**speed and stability optimizations** – Get the right answer for log(1+x) even when x is really tiny.
**dynamic C code generation** – Evaluate expressions faster.
**extensive unit-testing and self**-verification – Detect and diagnose many types of errors.

## 3. [TensorFlow](https://www.tensorflow.org/)
Similar to Theano, TensorFlow is an open source library for **numerical computation using data flow graphs** (which is all that a Neural Network really is). Nodes in the graph represent mathematical operations, while the graph edges represent the multidimensional data arrays (tensors) communicated between them. The flexible architecture allows you to deploy computation to one or more CPUs or GPUs in a desktop, server, or mobile device with a single API. TensorFlow was originally developed by researchers and engineers working on the Google Brain Team within Google's Machine Intelligence research organization for the purposes of conducting machine learning and deep neural networks research, but the system is general enough to be applicable in a wide variety of other domains as well. 
A primary benefit of TensorFlow (as compared to Theano) is **distributed computing, particularly among multiple-GPUs** (although this is something Theano is working on).

## 4. [Lasagne](https://github.com/Lasagne/Lasagne)
Lasagne is a lightweight library used to construct and train networks in Theano. The key term here is lightweight — it is not meant to be a heavy wrapper around Theano like Keras is. While this leads to your code being more verbose, it does free you from any restraints, while still giving you modular building blocks based on Theano.

- Supports feed-forward networks such as Convolutional Neural Networks (CNNs), recurrent networks including Long Short-Term Memory (LSTM), and any combination thereof
- Allows architectures of multiple inputs and multiple outputs, including auxiliary classifiers
- Many optimization methods including Nesterov momentum, RMSprop and ADAM
- Freely definable cost function and no need to derive gradients due to Theano's symbolic differentiation
- Transparent support of CPUs and GPUs due to Theano's expression compiler

Its design is governed by six principles:

- **Simplicity:** Be easy to use, easy to understand and easy to extend, to facilitate use in research
- **Transparency:** Do not hide Theano behind abstractions, directly process and return Theano expressions or Python / numpy data types
- **Modularity:** Allow all parts (layers, regularizers, optimizers, ...) to be used independently of Lasagne
- **Pragmatism:** Make common use cases easy, do not overrate uncommon cases
- **Restraint:** Do not obstruct users with features they decide not to use
- **Focus:** "Do one thing and do it well"

## 5. [Keras](https://keras.io/)
Keras is a high-level neural networks API, written in Python and capable of running on top of TensorFlow, CNTK, or Theano. It was developed with a focus on enabling fast experimentation. Being able to go from idea to result with the least possible delay is key to doing good research.
Use Keras if you need a deep learning library that:
- Allows for easy and fast **prototyping** (through user friendliness, modularity, and extensibility).
- Supports both convolutional networks and recurrent networks, as well as combinations of the two.
- Runs seamlessly on CPU and GPU.
Keras also places a heavy focus on Convolutional Neural Networks, this is extremely valuable from a computer vision perspective.
My only problem with Keras is that it **does not support multi-GPU environments** for training a network in parallel. This may or may not be a deal breaker for you.

## 6. [mxnet](https://github.com/apache/incubator-mxnet)
Apache MXNet (incubating) is a deep learning framework designed for both efficiency and flexibility. It allows you to **mix symbolic and imperative programming** to maximize efficiency and productivity. At its core, MXNet contains a dynamic dependency scheduler that automatically parallelizes both symbolic and imperative operations on the fly. A graph optimization layer on top of that makes symbolic execution fast and memory efficient. MXNet is **portable and lightweight**, **scaling effectively to multiple GPUs and multiple machines.**
While it can take a bit more code to standup a network in mxnet, what it does give you is an incredible number of language bindings (C++, Python, R, JavaScript, etc.) Again, it takes a little more code to get an experiment up and running in mxnet (as compared to Keras), but if you’re looking to distribute training across multiple GPUs or systems, I would use mxnet.

## 7. [sklearn-theano](https://github.com/sklearn-theano/sklearn-theano)
There are times where you **don’t need to train** a Convolutional Neural Network end-to-end. Instead, you need to treat the CNN as a **feature extractor**. This is especially useful in situations where you don’t have enough data to train a full CNN from scratch. Instead, just pass your input images through a popular pre-trained architecture such as OverFeat, AlexNet, VGGNet, or GoogLeNet, and extract features from the FC layers (or whichever layer you decide to use).

## 8. [nolearn](https://github.com/dnouri/nolearn)
While Keras wraps Theano and TensorFlow into a more user-friendly API, nolearn does the same — only for Lasagne. Furthermore, all code in nolearn is **compatible with scikit-learn**. Some don’t use nolearn for Convolutional Neural Networks (CNNs), although you certainly could (I prefer Keras and mxnet for CNNs) — I mainly use nolearn for its implementation of Deep Belief Networks (DBNs).

## 9. [DIGITS](https://github.com/NVIDIA/DIGITS)
DIGITS (the Deep Learning GPU Training System) is a **webapp** for training deep learning models. The currently supported frameworks are: **Caffe**, Torch, and Tensorflow. it can be quite tedious to define your _.prototxt_  files, generate your image dataset, run your network, and babysit your network training all via your terminal. DIGITS aims to fix this by allowing you to do (most) of these tasks in your browser. the user interface is excellent, providing you with valuable statistics and graphs as your model trains. I also like that you can easily visualize activation layers of the network for various inputs. Finally, if you have a specific image that you would like to test, you can either upload the image to your DIGITS server or enter the URL of the image and your Caffe model will automatically classify the image and display the result in your browser.

[Main Source](https://www.pyimagesearch.com/2016/06/27/my-top-9-favorite-python-deep-learning-libraries/)
